osquery/filesystem/filesystem.cpp

```diff
@@ -377,29 +383,52 @@ std::set<fs::path> getHomeDirectories() {
 bool safePermissions(const std::string& dir,
                      const std::string& path,
                      bool executable) {
-  struct stat file_stat, link_stat, dir_stat;
-  if (lstat(path.c_str(), &link_stat) < 0 || stat(path.c_str(), &file_stat) ||
-      stat(dir.c_str(), &dir_stat)) {
+  if (!platformIsFileAccessible(path).ok()) {
     // Path was not real, had too may links, or could not be accessed.
     return false;
   }
 
   if (FLAGS_allow_unsafe) {
     return true;
-  } else if (dir_stat.st_mode & (1 << 9)) {
+  }
+
+  Status result = platformIsTmpDir(dir);
+  if (!result.ok() && result.getCode() < 0) {
+    // An error has occurred in stat() on dir, most likely because the file path
+    // does not exist
+    return false;
+  } else if (result.ok()) {
     // Do not load modules from /tmp-like directories.
     return false;
-  } else if (S_ISDIR(file_stat.st_mode)) {
+  } 
+  
+  PlatformFile fd(path, PF_OPEN_EXISTING | PF_READ);
+  if (!fd.isValid()) {
+    return false;
+  }
+
+  result = fd.isDirectory();
+  if (!result.ok() && result.getCode() < 0) {
+    // Something went wrong when determining the file's directoriness
+    return false;
+  } else if (result.ok()) {
     // Only load file-like nodes (not directories).
     return false;
-  } else if (file_stat.st_uid == getuid() || file_stat.st_uid == 0) {
+  }
+
+  if (fd.isOwnerCurrentUser().ok() || fd.isOwnerRoot().ok()) {
+    result = fd.isExecutable();
+
     // Otherwise, require matching or root file ownership.
-    if (executable && !(file_stat.st_mode & S_IXUSR)) {
+    if (executable && ((!result.ok() && result.getCode() > 0) ||
```

Comment:
`Status::getCode` should imply `!=0` so if you want to check that it's `>0` you can leave out `!::ok()`.